server/server.go

```diff
@@ -887,34 +885,35 @@ func (s *server) getClusterInfo() *rpcpb.ClusterInfo {
 	return info
 }
 
-func (s *server) handleUnexpectedNodeStop(unexpectedNodeStopCh chan network.UnexpectedNodeStopMsg) {
+func (s *server) handleUnhealthy() {
+	timer := time.NewTimer(0)
+	if !timer.Stop() {
+		<-timer.C
+	}
+	defer timer.Stop()
+
 	for {
+		timer.Reset(healthCheckFreq)
 		select {
 		case <-s.closed:
 			return
 		case <-s.network.stopc:
 			return
 		case <-s.network.startErrc:
 			return
-		case unexpectedStopMsg := <-unexpectedNodeStopCh:
+		case <-timer.C:
+			ctx, cancel := context.WithTimeout(context.Background(), healthCheckTimeout)
+			err := s.network.nw.Healthy(ctx)
+			cancel()
+			if err == nil {
+				continue
+			}
 			zap.L().Info(
-				"received unexpected node stop message",
-				zap.String("node-name", unexpectedStopMsg.NodeName),
-				zap.Int("exit-code", unexpectedStopMsg.ExitCode),
+				"network is unhealthy",
```

Comment:
This actually means that the health call failed, not necessarily that it is unhealthy. It probably is in that case, but I am not sure about the semantics, would this check also have to fail if the network is just unhealthy?

Reply:
I'm not sure I follow. This is in the server, not the client. `Healthy` returns a non-nil error if a node isn't healthy.